### PR TITLE
Split compile tests into smaller chunks to improve CI times

### PR DESCRIPTION
**Pull request by serban-nicusor-toptal**
#### Submission Checklist

- [ ] Run unit tests
- Documentation
    - [ ] If a user-facing facing change was made, the documentation PR is here: <LINK>
    - [ ] OR, no user-facing changes were made

## Release notes

Improve CI builds times by splitting compile tests and running them in parallel.

## Copyright and Licensing

By submitting this pull request, the copyright holder is agreeing to 
license the submitted work under the BSD 3-clause license (https://opensource.org/licenses/BSD-3-Clause)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -3,9 +3,9 @@ This repo contains a new compiler for Stan, stanc3, written in OCaml.
 Since version 2.26, this has been the default compiler for Stan. See [this wiki](https://github.com/stan-dev/stanc3/wiki/changes-from-stanc2) for a list of minor differences between this compiler and the previous Stan compiler.
 
 To read more about why we built this, see this [introductory blog post](https://statmodeling.stat.columbia.edu/2019/03/13/stanc3-rewriting-the-stan-compiler/). For some discussion as to how we chose OCaml, see [this accidental flamewar](https://discourse.mc-stan.org/t/choosing-the-new-stan-compilers-implementation-language/6203).
-We're testing [these models](https://jenkins.mc-stan.org/job/stanc3/job/master/) (listed under Test Results) on every pull request.
+We're testing [these models](https://jenkins.flatironinstitute.org/job/stanc3/job/master/) (listed under Test Results) on every pull request.
 
-[![Build Status](https://jenkins.mc-stan.org/buildStatus/icon?job=stanc3%2Fmaster&style=flat-square)](https://jenkins.mc-stan.org/job/stanc3/job/master/)
+[![Build Status](https://jenkins.flatironinstitute.org/buildStatus/icon?job=stanc3%2Fmaster&style=flat-square)](https://jenkins.flatironinstitute.org/job/stanc3/job/master/)
```
Comment (WardBrian) [inline README.md:8]: These links are 404s (I think they're missing the extra Stan folder?)

**Comment by WardBrian** [comment]
This seems like a reasonable place to start. If we need to we can also divide up the tests inside /good - I belive the majority of them end up being inside good/function-signatures

**Comment by serban-nicusor-toptal** [comment]
- performance-tests-cmdstan/example-models
```
.:	1746

./jupyter:	10
./jupyter/covid-inf-rate:	9

./education:	44
./education/rasch_and_2pl:	2
./education/tutorial_twopl:	10
./education/rsm_and_grsm:	2
./education/tutorial_rstanarm:	2
./education/causal_rct:	9
./education/dina_independent:	8
./education/hierarchical_2pl:	4
./education/pcm_and_gpcm:	2

./basic_distributions:	11
./applications:	3
./applications/covid-19:	3

./knitr:	362
./knitr/reduce-sum:	6
./knitr/sum-of-exponentials:	3
./knitr/car-iar-poisson:	68
./knitr/planetary_motion:	10
./knitr/disease_transmission:	8
./knitr/irt:	32
./knitr/convert-odes:	10
./knitr/pedantic:	1
./knitr/cloud-compute-2020:	26
./knitr/pool-binary-trials:	19
./knitr/hmm-example:	2
./knitr/chapter2:	32
./knitr/cook_et_al:	8
./knitr/bandits:	7
./knitr/sex-ratio:	2
./knitr/lotka-volterra:	8
./knitr/golf:	17
./knitr/single-exponential:	4
./knitr/world-cup:	14
./knitr/mle-params:	6
./knitr/curse-dims:	2
./knitr/soil-carbon:	8
./knitr/bradley-terry:	7
./knitr/climate-challenge:	4
./knitr/simplest-regression:	6
./knitr/pest-control:	17
./knitr/ratings:	4
./knitr/dorazio-royle-occupancy:	4
./knitr/neural-nets:	13
./knitr/movement-hmm:	1
./knitr/chapter1:	7
./knitr/noisy-raters:	6

./Bayesian_Cognitive_Modeling:	120
./Bayesian_Cognitive_Modeling/CaseStudies:	52
./Bayesian_Cognitive_Modeling/GettingStarted:	3
./Bayesian_Cognitive_Modeling/ModelSelection:	8
./Bayesian_Cognitive_Modeling/! other:	18
./Bayesian_Cognitive_Modeling/ParameterEstimation:	38

./basic_estimators:	20
./regression_tests:	4

./BPA:	185
./BPA/Ch.09:	20
./BPA/Ch.12:	15
./BPA/Ch.07:	43
./BPA/Ch.10:	13
./BPA/Ch.03:	6
./BPA/Ch.11:	13
./BPA/Ch.04:	17
./BPA/Ch.06:	19
./BPA/Ch.13:	20
./BPA/Ch.05:	5
./BPA/Ch.08:	12

./bugs_examples:	267
./bugs_examples/vol3:	28
./bugs_examples/vol2:	112
./bugs_examples/vol1:	126

./misc:	155
./misc/hier_multivariate:	2
./misc/irt:	13
./misc/multi-logit:	6
./misc/nnmf:	7
./misc/ecology:	12
./misc/cluster:	13
./misc/eight_schools:	2
./misc/gam:	3
./misc/funnel:	2
./misc/gaussian-process:	34
./misc/linear-regression:	2
./misc/dlm:	11
./misc/garch:	8
./misc/multivariate-probit:	5
./misc/sur:	4
./misc/hmm:	13
./misc/moving-avg:	18

./ARM:	557
./ARM/Ch.21:	26
./ARM/Ch.2:	5
./ARM/Ch.5:	49
./ARM/Ch.22:	11
./ARM/Ch.7:	65
./ARM/Ch.12:	23
./ARM/Ch.24:	9
./ARM/Ch.25:	13
./ARM/Ch.23:	32
./ARM/Ch.18:	17
./ARM/Ch.3:	24
./ARM/Ch.15:	1
./ARM/Ch.19:	30
./ARM/Ch.9:	20
./ARM/Ch.16:	22
./ARM/Ch.20:	11
./ARM/Ch.10:	26
./ARM/Ch.11:	2
./ARM/Ch.6:	15
./ARM/Ch.17:	34
./ARM/Ch.13:	32
./ARM/Ch.4:	54
./ARM/Ch.8:	19
./ARM/Ch.14:	16

```

stanc3/tests/integration/good:

```
.:	1082

./indexing:	10

./code-gen:	56
./code-gen/ode:	4
./code-gen/standalone_functions:	5
./code-gen/expressions:	4
./code-gen/opencl:	4
./code-gen/profiling:	4

./services:	8
./compound-assign:	30
./ode:	11

./compiler-optimizations:	47
./compiler-optimizations/mem_patterns:	11

./promotion:	10
./warning:	38
./mcmc:	15
./mcmc/hmc:	13
./array-expr:	8
./for_loops:	5
./vec-expr:	7
./variational:	14
./model:	4
./lang:	33

./function-signatures:	553
./function-signatures/math:	286
./function-signatures/distributions:	266

./overloading:	14
./runtime_errors:	8
./stat_comp_benchmarks_models:	16
./parser-generator:	34
./optimization:	4
```

Above is a list and count, what do you think should I make a bash/python script to split them during the CI or just do it in github dir structure ?

**Comment by WardBrian** [comment]
It seems like you'll need to touch a file in `src` to really test this PR, I think. We might want to add a build parameter to override those checks and run the full pipeline?

**Comment by serban-nicusor-toptal** [comment]
We have a manual step in place, if we retry the job it will run all the steps.
But I agree that we should have something a bit more automated, we can read the commit message and if it contains something like 'ci-skip' or 'verify-skip' it'll just run everything.

**Comment by WardBrian** [comment]
I propose something like [ci skip] for skipping and [ci run all] for running the entire pipeline regardless of what verifyChanges says. If that's possible

File: README.md (hunk 1 above)
Comment (serban-nicusor-toptal) [reply README.md]: Should be fixed now, poor editing from my side.

**Comment by WardBrian** [comment]
It’s fascinating to me that even though example models has so many more models it’s so much quicker. I think we will probably want to do both this split and reduce how often we run the distributions in the function-signatures folder

**Comment by serban-nicusor-toptal** [comment]
Hey @WardBrian [here's the PR](https://github.com/stan-dev/jenkins-shared-libraries/pull/4) about [ci skip], would you mind taking a look ? Sadly with Jenkins I can't really give it a success message when you want to [ci skip].

**Comment by WardBrian** [comment]
Would it be reasonable to treat those tests like the math expression ones and only run them in a PR if a signature was added or dropped, and then also run them on master or weekly/monthly?

**Comment by serban-nicusor-toptal** [comment]
I just saw that `-j` is set to `8` still, here it's 4 ( we have 4 cores allocated / executor ) maybe this slows it down that much ?

**Comment by rok-cesnovar** [comment]
Will try it out in the other PR, lets see.

**Comment by WardBrian** [comment]
```
                         echo 'O=0' >> make/local
                            echo 'CXX=${LINUX_CXX}' >> make/local
                            echo 'PRECOMPILED_HEADERS=false' >> make/local
                            make clean-all
                            make -j${PARALLEL} build

```

vs


```
                            cd performance-tests-cmdstan
                                mkdir cmdstan/bin
                                cp ../bin/stanc cmdstan/bin/linux-stanc
                                cd cmdstan; make clean-all; make -j${env.PARALLEL} build
```

Seems like maybe we should try setting those make variables?

**Comment by rok-cesnovar** [comment]
Ah, OK, the first thing is we compile with `-O3` here, which we don't need to do. These integration tests were always run with O=0, which I though https://github.com/stan-dev/stanc3/blob/19be33f3a6b357310bb10a64b960ce5087cc9e01/Jenkinsfile#L200 should take care off, but seems not to be the case.

**Comment by WardBrian** [comment]
@rok-cesnovar we should also set O=0 in the cmdstan instance used for the math expression tests that only run when signatures change

**Comment by rok-cesnovar** [comment]
That is there: https://github.com/stan-dev/stanc3/blob/ba71588babb320c3167433870d656b5a1aa405d7/Jenkinsfile#L315

Added O=0 here now. Hopefully this settles it for some time.

**Comment by WardBrian** [comment]
Thanks @rok-cesnovar!

**Comment by WardBrian** [comment]
The model end to end tests are still using `O3`, but I guess that makes sense for a performance test?

**Comment by rok-cesnovar** [comment]
Yeah, I think those are intentionally O3, it's also not compiling a ton of models there.

File: Jenkinsfile [hunk 2]
```diff
@@ -1,7 +1,7 @@
-@Library('StanUtils')
+@Library('StanUtils@add-ci-skip')
```
Comment (rok-cesnovar) [inline Jenkinsfile:1]: Does this have to be cleaned up?

Comment (serban-nicusor-toptal) [reply Jenkinsfile:1]: Yes

**Comment by serban-nicusor-toptal** [comment]
Wouldn't `-j8` make it worse when we're at 4 cores ?

**Comment by rok-cesnovar** [comment]
I am just trying to test and see. It might be worse, but who knows :) Worth a try. For -j4 and all the changes here, the build times were 17min for example models and 45 for the rest. 

Will revert this change if it proves to be worse. Other than the cleanup with the StanUtils this then looks good to go!

**Comment by rok-cesnovar** [comment]
We dont gain anything with j8, so 4 it is.

**Review by rok-cesnovar — APPROVED** [review]
Thanks Nic!

**Comment by serban-nicusor-toptal** [comment]
Do we split the `good` tests even more ? Right now it takes about 45 minutes

**Comment by rok-cesnovar** [comment]
I think 45 min is good enough for now.

**Comment by WardBrian** [comment]
Agreed, we can do more later if needed

**Pull request merged**